internal/database/database.go

```diff
@@ -0,0 +1,150 @@
+package database
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"github.com/atburke/teleport_interview/internal/crypto"
+	"github.com/atburke/teleport_interview/internal/types"
+	_ "github.com/go-sql-driver/mysql"
+	"time"
+)
+
+// tokens and password hash could be []byte objects, but they're strings in the
+// db and in the api, so leaving them as hex strings will save time
+
+// Database represents a database connection.
+type Database interface {
+
+	// CreatePreAuthSession creates a new session for a user that has not yet
+	// authenticated.
+	CreatePreAuthSession(csrfToken string, initTime time.Time) (*types.Session, error)
+
+	// CreateSession creates a new session for an authenticated user.
+	CreateSession(accountId, csrfToken string, initTime time.Time) (*types.Session, error)
+
+	// FetchSession fetches the session identified by the provided token.
+	FetchSession(sessionToken string) (*types.Session, error)
+
+	// DeleteSession deletes a session for a user.
+	DeleteSession(sessionToken string) error
+
+	// FetchAccount fetches the account for a user.
+	FetchAccount(email string) (*types.Account, error)
+
+	// DeleteExpiredSessions deletes sessions that have expired (after either their
+	// idle or absolute timeout has passed)
+	DeleteExpiredSessions(t time.Time) error
+
+	// Close closes any underlying resources.
+	Close()
```

Comment:
Curious why doesn't this return an error? If it did, it'd make it compatible with `io.Closer`.